WcaOnRails/lib/advance_to_next_round_requirement.rb

```diff
@@ -27,21 +28,101 @@ def hash
     self.to_wcif.hash
   end
 
+  def self.wcif_type_to_class
+    @@wcif_type_to_class ||= AdvanceToNextRoundRequirement.subclasses.map { |cls| [cls.wcif_type, cls] }.to_h
+  end
+
   def self.load(json)
     if json.nil? || json.is_a?(self)
       json
     else
-      self.new.tap do |cutoff|
-        json_obj = json.is_a?(Hash) ? json : JSON.parse(json)
-        cutoff.type = json_obj['type']
-        cutoff.ranking = json_obj['ranking']
-        cutoff.percentile = json_obj['percentile']
-        cutoff.attemptValue = json_obj['attemptValue']
-      end
+      json_obj = json.is_a?(Hash) ? json.clone : JSON.parse(json)
+      wcif_type = json_obj.delete('type')
+      self.wcif_type_to_class[wcif_type].new(json_obj.symbolize_keys)
     end
   end
 
   def self.dump(cutoff)
     cutoff ? JSON.dump(cutoff.to_wcif) : nil
   end
 end
+
+class RankingRequirement < AdvanceToNextRoundRequirement
+  attr_accessor :ranking
+  validates :ranking, numericality: { only_integer: true }
+
+  def initialize(ranking:)
+    self.ranking = ranking
+  end
+
+  def self.wcif_type
+    "ranking"
+  end
+
+  def value
+    ranking
+  end
+
+  def to_s(round)
+    next_round_number = round.number + 1
+    I18n.t("advance_to_next_round_requirement.ranking", ranking: ranking, next_round_number: next_round_number)
+  end
+end
+
+class PercentileRequirement < AdvanceToNextRoundRequirement
+  attr_accessor :percentile
+  validates :percentile, numericality: { only_integer: true }
+
+  def initialize(percentile:)
+    self.percentile = percentile
+  end
+
+  def self.wcif_type
+    "percentile"
+  end
+
+  def value
+    percentile
+  end
+
+  def to_s(round)
+    next_round_number = round.number + 1
+    I18n.t("advance_to_next_round_requirement.percentile", percentile: percentile, next_round_number: next_round_number)
+  end
+end
+
+class AttemptValueRequirement < AdvanceToNextRoundRequirement
+  attr_accessor :attempt_value
+  validates :attempt_value, numericality: { only_integer: true }
+
+  def initialize(attempt_value: nil, attemptValue: nil)
```

Comment:
We store camelCase in the database, and this constructor is called by the deserialization code in the load method, which means an attemptValue is passed in. However, there's also Ruby test code that passes in `attempt_value`, so we handle both of those here.

An alternative would be to store snake_case in the database, and convert to / from camelCase whenever dealig with WCIF data. That would require creating 2 kinds of load methods and 2 kinds of serialize methods, though (1 for snake_case and 1 for camelCase).